projects/python-client/src/datapane/client/api/user.py

```diff
@@ -113,6 +116,76 @@ def _run_script(script: str):
     runpy.run_path(str(script_path), run_name="__datapane__")
 
 
+def _run_template(template_path: str):
+    """Run the template script"""
+    # Store the cwd before we change it
+    old_cwd = os.getcwd()
+    # Change cwd to support relative paths in template script
+    os.chdir(template_path)
+
+    try:
+        runpy.run_path(
+            "template.py",
+            run_name="__datapane__",
+        )
+    # Notify the user of missing packages that are required by the template
+    except ModuleNotFoundError as e:
+        raise DPError(f"Please install the following packages to run this template\n{e.name}") from e
+
+    os.chdir(old_cwd)
+
+
+def _download_template(url: str):
+    """Download the template from a repository url, and delete the .git directory
+
+    Returns:
+        The path of the cloned template repository
+    """
+
+    url = _check_repo_url(url)
+
+    # Check if target directory already exists
+    target = url.split("/")[-1]
+    target_dir_exists = os.path.exists(target)
+
+    # If directory exists, raise an error notifying the user
+    if target_dir_exists:
+        raise DPError(f"Directory {target} already exists.")
+
+    # Shallowest clone of the template repo
+    template_repo = porcelain.clone(url, target=target, depth=1)
+
+    # Remove .git directory
+    shutil.rmtree(template_repo.controldir())
+
+    return template_repo.path
+
+
+def _check_repo_url(url: str):
+    """Check if the template repository exists. Supports first or third-party templates.
+
+    Returns:
+        The absolute uri of the template repository.
+    """
+    # Check if remote is a git repo
+    try:
+        porcelain.ls_remote(url)
+    except (errors.NotGitRepository, client.HTTPUnauthorized):
+        try:
+            # Try appending the supplied url to the datapane organization
+            full_url = f"https://github.com/datapane/{url}"
```

Comment:
While I support the ask-for-forgiveness approach in general, I think in this case we should check the URL doesn't have special characters in (`:` and `/` should be enough).

This avoids some unknowns, like someone trying to reference a local directory incorrectly and accidentally downloading + executing a strange repo.

```
poetry run datapane template ../antdking/clean-actions
Retrieving and running the template at `../antdking/clean-actions`.  <-- fetched from github

Enumerating objects: 83, done.
Counting objects: 100% (83/83), done.
```

Reply:
Added check for `:` and `/` before appending to datapane organization